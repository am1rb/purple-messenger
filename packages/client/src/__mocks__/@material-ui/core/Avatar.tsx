import React from "react";
import type { AvatarProps } from "@material-ui/core/Avatar";

function MuiAvatar({ src, alt }: AvatarProps) {
  return <img data-testid="mock-mui-avatar" src={src} alt={alt} />;
}

export default MuiAvatar;
