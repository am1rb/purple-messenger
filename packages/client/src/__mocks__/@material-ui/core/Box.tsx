import React from "react";

export interface BoxProps {
  children: React.ReactNode;
}

const muiBox = jest.fn(({ children }: BoxProps) => (
  <div data-testid="mui-box">{children}</div>
));

export default muiBox;
