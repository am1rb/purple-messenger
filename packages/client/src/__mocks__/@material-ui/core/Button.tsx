import React from "react";

export interface ButtonProps {
  onClick?: () => void;
  children: React.ReactNode;
}

function MuiButton({ onClick, children }: ButtonProps) {
  return (
    <div data-testid="mock-mui-button" onClick={onClick}>
      {children}
    </div>
  );
}

export default MuiButton;
