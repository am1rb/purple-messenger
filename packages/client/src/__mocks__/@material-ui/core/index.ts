export { default as Avatar } from "./Avatar";
export { default as Box } from "./Box";
export { default as Card } from "./Card";
export { default as CardContent } from "./CardContent";
export { default as Typography } from "./Typography";
export { default as Button } from "./Button";
