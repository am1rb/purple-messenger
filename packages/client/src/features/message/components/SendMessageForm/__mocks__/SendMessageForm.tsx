import React from "react";

export interface SendMessageFormProps {
  disabled: boolean;
}

function SendMessageForm({ disabled }: SendMessageFormProps) {
  return <form data-testid="mock-send-message-form" data-disabled={disabled} />;
}

export default SendMessageForm;
