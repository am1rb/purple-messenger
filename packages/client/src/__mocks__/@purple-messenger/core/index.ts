import * as purpleMessengerCore from "@purple-messenger/core";

const mockedPurpleMessengerCore = jest.genMockFromModule<
  typeof purpleMessengerCore
>("@purple-messenger/core");

function mockedAction(action = () => ({ type: "" })) {
  return jest.fn(action);
}

module.exports = {
  ...mockedPurpleMessengerCore,
  loadConversationList: mockedAction(),
  unloadConversationList: mockedAction(),
  signOut: mockedAction(),
  clearMessageList: mockedAction(),
  seenMessageAck: mockedAction(),
  start: mockedAction(),
  shutdown: mockedAction(),
  submitMessage: mockedAction(),
  startTypingMessage: mockedAction(),
  stopTypingMessage: mockedAction(),
  signIn: mockedAction(),
  setAuthError: mockedAction(),
  verifyToken: mockedAction(),
  connected: mockedAction(purpleMessengerCore.connected),
  disconnected: mockedAction(purpleMessengerCore.disconnected),
};
