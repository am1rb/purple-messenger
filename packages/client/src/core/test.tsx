import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { ConnectedRouter } from "connected-react-router";
import configureStore, { history } from "core/redux/configureStore";
import { State as RootState } from "core/redux/reducers";
import theme from "core/style/theme";
import { ThemeProvider } from "@material-ui/styles";

export function renderWithStore(
  ui: React.ReactElement,
  initStore: Partial<RootState>
) {
  const store = configureStore(initStore);

  return render(ui, {
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <ConnectedRouter history={history}>{children}</ConnectedRouter>
        </Provider>
      </ThemeProvider>
    ),
  });
}
