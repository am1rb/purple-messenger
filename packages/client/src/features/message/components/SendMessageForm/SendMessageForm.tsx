import React, { useCallback, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { SubmitHandler } from "@rocketseat/unform";
import { Box, IconButton } from "@material-ui/core";
import SendIcon from "@material-ui/icons/Send";
import {
  startTypingMessage,
  stopTypingMessage,
  submitMessage,
} from "@purple-messenger/core";
import Unform from "components/Unform";
import MessageTextField from "../MessageTextField";
import useConversation from "features/conversation/components/useConversationInfo";
import { MessageFormContent } from "features/message/types/formContent";

export interface SendMessageFormProps {
  disabled: boolean;
}

function SendMessageForm({ disabled }: SendMessageFormProps) {
  const dispatch = useDispatch();
  const { username } = useConversation();
  const buttonRef = useRef<HTMLButtonElement>(null);

  const handleSubmit = useCallback<SubmitHandler<MessageFormContent>>(
    (data, helpers) => {
      if (username) {
        dispatch(submitMessage(username, data.message));
      }
      helpers.resetForm();
    },
    [dispatch, username]
  );

  const handleEnter = useCallback(() => buttonRef.current?.click(), []);

  const handleStartTyping = useCallback(() => {
    if (username) {
      dispatch(startTypingMessage(username));
    }
  }, [dispatch, username]);

  const handleStopTyping = useCallback(() => {
    if (username) {
      dispatch(stopTypingMessage(username));
    }
  }, [dispatch, username]);

  return (
    <Unform<MessageFormContent> onSubmit={handleSubmit}>
      <Box display="flex" alignItems="center">
        <Box flexShrink="0" flexGrow="1" mr={1}>
          <MessageTextField
            name="message"
            disabled={disabled}
            onEnter={handleEnter}
            onStartTyping={handleStartTyping}
            onStopTyping={handleStopTyping}
            autoFocus
          />
        </Box>
        <Box flexShrink="0" flexGrow="0">
          <IconButton type="submit" disabled={disabled} ref={buttonRef}>
            <SendIcon />
          </IconButton>
        </Box>
      </Box>
    </Unform>
  );
}

export default SendMessageForm;
