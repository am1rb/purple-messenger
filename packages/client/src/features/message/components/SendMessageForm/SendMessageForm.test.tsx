import React from "react";
import SendMessageForm, { SendMessageFormProps } from "./SendMessageForm";
import { renderWithStore } from "core/test";
import userEvent from "@testing-library/user-event";
import { fireEvent } from "@testing-library/react";
import {
  startTypingMessage,
  stopTypingMessage,
  submitMessage,
} from "@purple-messenger/core";

jest.mock("components/Unform");
jest.mock("../MessageTextField");

const sharedProps: SendMessageFormProps = {
  disabled: false,
  username: "john",
};
const sharedMessage = "Hello World!";

describe("The <SendMessageForm /> tests", () => {
  it("Should match the snapshot", () => {
    const { container } = renderWithStore(<SendMessageForm {...sharedProps} />);
    expect(container).toMatchSnapshot();
  });

  it("Should send the message by clicking on the send button", () => {
    const message = "Hello World!";
    const { getByTestId } = renderWithStore(
      <SendMessageForm {...sharedProps} />
    );

    userEvent.type(getByTestId("mock-message-text-field"), message);
    userEvent.click(getByTestId("mock-mui-icon-button"));

    expect(submitMessage).toHaveBeenCalledWith(sharedProps.username, message);
  });

  it("Should send the message by pressing the enter button", () => {
    const { getByTestId } = renderWithStore(
      <SendMessageForm {...sharedProps} />
    );

    const messageField = getByTestId("mock-message-text-field");
    userEvent.type(messageField, sharedMessage);
    fireEvent.keyDown(messageField, { keyCode: 13 });

    expect(submitMessage).toHaveBeenCalledWith(
      sharedProps.username,
      sharedMessage
    );
  });

  it("Should call startTypingMessage when something is typed", () => {
    const { getByTestId } = renderWithStore(
      <SendMessageForm {...sharedProps} />
    );

    const messageField = getByTestId("mock-message-text-field");
    userEvent.type(messageField, sharedMessage);

    expect(startTypingMessage).toHaveBeenCalledWith(sharedProps.username);
  });

  it("Should call stopTypingMessage when something is typed", () => {
    jest.useFakeTimers();

    const { getByTestId } = renderWithStore(
      <SendMessageForm {...sharedProps} />
    );

    const messageField = getByTestId("mock-message-text-field");
    userEvent.type(messageField, sharedMessage);

    jest.runAllTimers();

    expect(stopTypingMessage).toHaveBeenCalledWith(sharedProps.username);

    jest.clearAllTimers();
  });

  it("Should disable the text input and the button if the disable prop is passed", () => {
    const { getByTestId } = renderWithStore(
      <SendMessageForm {...sharedProps} disabled />
    );
    expect(getByTestId("mock-message-text-field")).toHaveAttribute("disabled");
    expect(getByTestId("mock-mui-icon-button")).toHaveAttribute("disabled");
  });
});
