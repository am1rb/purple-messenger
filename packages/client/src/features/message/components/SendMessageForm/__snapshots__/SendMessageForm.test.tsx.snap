// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <SendMessageForm /> tests Should match the snapshot 1`] = `
<div>
  <form>
    <div
      data-testid="mock-mui-box"
    >
      <div
        data-testid="mock-mui-box"
      >
        <input
          data-testid="mock-message-text-field"
          name="message"
        />
      </div>
      <div
        data-testid="mock-mui-box"
      >
        <button
          data-testid="mock-mui-icon-button"
          type="submit"
        >
          <span
            data-testid="mock-mui-icons-send"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;
